### PR TITLE
Tweak open_dataset to ingest multiple variables at once

### PR DESCRIPTION
**Pull request by htmlboss**
Resolves T/S plot crashes.

When the sql back-end was implemented, `open_dataset` didn't consider multiple variables in the `variable` kwarg. This resulted in T/S plots failing for datasets that had variables split across multiple files. This PR fixes that.

## Background

* Also updated a netcdf fixture to re-enable some API tests.

## Why did you take this approach?
Path of least resistance. This is actually finishing a part of the sqlite refactor I started back in Aug 2019. Forgot to finish it when Fall came around.

## Anything in particular that should be highlighted?
More tests have been enabled!

## Screenshot(s)
image

## Checks
- [x] I ran unit tests.
- [x] I've tested the relevant changes from a user POV.

_Hint_ To run all python unit tests run the following command from the root repository directory:
```sh
python -m unittest `find tests -name "*.py" | grep -v disabled`
```

### CODE REVIEW

File: data/netcdf_data.py [hunk 1]
```diff
@@ -757,34 +757,46 @@ def get_nc_file_list(self, datasetconfig: DatasetConfig, **kwargs: dict) -> Unio
...
+            if not isinstance(variable, set):
+                variable = set(variable)
 
-            calculated_variables = datasetconfig.calculated_variables
-            if variable[0] in calculated_variables:
-                equation = calculated_variables[variable[0]]['equation']
-
-                variable = data.utils.get_data_vars_from_equation(
-                    equation, [v.key for v in db.get_data_variables()])
+            raise
```
Comment (htmlboss) [inline data/netcdf_data.py]: temporary raise for debugging. will be deleted before requesting review

File: data/netcdf_data.py [hunk 2]
```diff
@@ -757,34 +757,46 @@ def get_nc_file_list(self, datasetconfig: DatasetConfig, **kwargs: dict) -> Unio
...
-                variable = [variable]
+            if not isinstance(variable, set):
+                variable = set(variable)
 
-            calculated_variables = datasetconfig.calculated_variables
-            if variable[0] in calculated_variables:
-                equation = calculated_variables[variable[0]]['equation']
-
-                variable = data.utils.get_data_vars_from_equation(
-                    equation, [v.key for v in db.get_data_variables()])
```
Comment (douglatornell) [inline data/netcdf_data.py]: I stumbled across this issue a few days ago when I noticed that CTD profiles and T/S diagrams weren't working. Sorry, I didn't mention it, or provide a patch, but here is the solution I came up with:
```python
            if not isinstance(variable, list):
                variable = set([variable])
            else:
                variable = set(variable)

            calculated_variables = datasetconfig.calculated_variables
            for var in variable:
                if var in calculated_variables:
                    equation = calculated_variables[var]['equation']
                    equation_vars = data.utils.get_data_vars_from_equation(
                            equation, [v.key for v in db.get_data_variables()])
                    variable.union(set(equation_vars))
            variable = list(variable)
```
More casting back and forth between `list` and `set` than I would like - perhaps some can be polished. Using `set` to avoid duplication of vars from eqns, IIRC.

Grist for the mill...

Comment (htmlboss) [reply data/netcdf_data.py]: No worries! It's a reassurance for me that my solution is similar in thinking to yours :1st_place_medal:
To avoid the `variable.union(set(equation_vars))` cast,  I used a `variable.update(equation_vars)` which grabs stuff from other iterables.

File: data/netcdf_data.py [hunk 3]
```diff
@@ -757,32 +757,42 @@ def get_nc_file_list(self, datasetconfig: DatasetConfig, **kwargs: dict) -> Unio
...
+            variables_to_load = self.__get_variables_to_load(db, variable, calculated_variables)
 
             try:
                 timestamp = self.__get_requested_timestamps(
-                    db, variable[0], kwargs['timestamp'], kwargs.get('endtime'), kwargs.get('nearest_timestamp', False))
+                    db, variables_to_load[0], kwargs['timestamp'], kwargs.get('endtime'), kwargs.get('nearest_timestamp', False))
```
Comment (htmlboss) [inline data/netcdf_data.py]: the timestamp and endtime kwargs are the same for all variables when loading, so just grab the first variable.

File: data/netcdf_data.py [hunk 4]
```diff
@@ -757,32 +757,42 @@ def get_nc_file_list(self, datasetconfig: DatasetConfig, **kwargs: dict) -> Unio
...
             if not file_list:
                 raise RuntimeError("NetCDF file list is empty.")
 
             self._nc_files = file_list
+    
+    def __get_variables_to_load(self, db: SQLiteDatabase, variable: set,
+                                    calculated_variables: dict) -> List[str]:
+        
+        calc_var_keys = set(calculated_variables)
+        variables_to_load = variable.difference(calc_var_keys)
```
Comment (htmlboss) [inline data/netcdf_data.py]: I could do `variable - calc_var_keys` here but I think `.difference` is more clear.

File: data/sqlite_database.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ def __exit__(self, exc_type, exc_value, traceback):
     def __flatten_list(self, some_list: list) -> list:
         return list(itertools.chain(*some_list))
 
-    def get_netcdf_files(self, timestamp: list, variable: list) -> List[str]:
+    def get_netcdf_files(self, timestamp: List[int], variable: List[str]) -> List[str]:
```
Comment (htmlboss) [inline data/sqlite_database.py]: drive-by type hint improvement.

**Comment by jamesprayner** [comment]
Have you tried running the tests with the new subset file AFTER clearing your navigator cache? Just want to make sure they all still work.

**Comment by htmlboss** [comment]
> Have you tried running the tests with the new subset file AFTER clearing your navigator cache? Just want to make sure they all still work.

Good catch. Just tried it and no additional problems.
